pkg/cachedinstances/cloudprovider/cached_cloud_provider.go

```diff
@@ -0,0 +1,261 @@
+	statsCacheRefreshNegative uint64 // Cumulative number of negative refreshes (ie, a refresh which failed and used old data)
+	statsCachePositive        uint64 // Absolute number of positive entries in cache
+	statsCacheNegative        uint64 // Absolute number of negative entries in cache
+
+	logger         logrus.FieldLogger
+	limiter        *rate.Limiter
+	cloudProvider  gostatsd.CloudProvider
+	cacheOpts      gostatsd.CacheOptions
+	ipSinkSource   chan gostatsd.IP
+	infoSinkSource chan gostatsd.InstanceInfo
+
+	// emitChan triggers a write of all the current stats when it is given a Statser
+	emitChan     chan stats.Statser
+	rw           sync.RWMutex // Protects cache
+	cache        map[gostatsd.IP]*instanceHolder
+	toLookupIPs  []gostatsd.IP
+	toReturnInfo []gostatsd.InstanceInfo
+}
+
+func (ch *CachedCloudProvider) Run(ctx context.Context) {
+	var (
+		toLookupC     chan<- gostatsd.IP
+		toLookupIP    gostatsd.IP
+		toReturnInfoC chan<- gostatsd.InstanceInfo
+		toReturnInfo  gostatsd.InstanceInfo
+		wg            wait.Group
+	)
+	// this goroutine needs to populate/update the cache so an intermediate InstanceInfo channel is used below that allows
+	// to intercept, update the cache and then push the information through to the cache consumer.
+	ownInfoSource := make(chan gostatsd.InstanceInfo)
+	ld := cloudProviderLookupDispatcher{
+		logger:        ch.logger,
+		limiter:       ch.limiter,
+		cloudProvider: ch.cloudProvider,
+		ipSource:      ch.ipSinkSource, // our sink is their source
+		infoSink:      ownInfoSource,   // their sink is our source
+	}
+
+	defer wg.Wait() // Wait for cloudProviderLookupDispatcher to stop
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel() // Tell cloudProviderLookupDispatcher to stop
+
+	wg.StartWithContext(ctx, ld.run)
+
+	refreshTicker := time.NewTicker(ch.cacheOpts.CacheRefreshPeriod)
+	defer refreshTicker.Stop()
+	// No locking for ch.cache READ access required - this goroutine owns the object and only it mutates it.
+	// So reads from the same goroutine are always safe (no concurrent mutations).
+	// When we mutate the cache, we hold the exclusive (write) lock to avoid concurrent reads.
+	// When we read from the cache from other goroutines in the Peek() method, we obtain the read lock.
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case toLookupC <- toLookupIP:
+			toLookupIP = gostatsd.UnknownIP // enable GC
+			toLookupC = nil                 // ip has been sent; if there is nothing to send, the case is disabled
+		case toReturnInfoC <- toReturnInfo:
+			toReturnInfo = gostatsd.InstanceInfo{} // enable GC
+			toReturnInfoC = nil                    // info has been sent; if there is nothing to send, the case is disabled
+		case info := <-ownInfoSource:
+			ch.handleInstanceInfo(info)
+		case t := <-refreshTicker.C:
+			ch.doRefresh(t)
+		case statser := <-ch.emitChan:
+			ch.emit(statser)
+		}
+		if toLookupC == nil && len(ch.toLookupIPs) > 0 {
```

Comment:
`ch.toLookupIPs` is only ever populated by `ch.doRefresh()`.  `ch.doRefresh()` could just return a slice of IPs to refresh, then this could pass that entire thing as a single slice to `ch.ipSinkSource`, rather than feeding it through one at a time.

Reply:
What are the benefits? This will complicate logic quite a bit. `cloudProviderLookupDispatcher.run()` would need to be handle all the extra complexity of collecting just enough IPs in a buffer and then processing them. I tried such approach somewhere during this refactoring and found that it complicates things, hurts readability. And when things are not simple, bugs creep in. :)

Reply:
Splitting this in to two aspects, which should be considered individually - returning a slice, and passing a slice to the lookup.

Returning a slice removes the extra state from the structure, simplifying the query of "where is this referenced?". Reducing the scope to a local variable means I don't need to have to think about locking, or what else might touch it, because I can see it's a regular local variable, returned from a function.  The only thing that it simplifies currently is no back pressure. If back pressure were added so we didn't call `doRefresh`, then it wouldn't even add that, we'd just be returning a slice through the struct.

Passing a slice to lookup means the lookup dispatcher has to do a little extra thinking, but I wouldn't say it's much. Without writing code, I'd probably wrap it up in a type which has 3 methods... "push more", "pop maxLookupIps or nil", and "get maxLookupIPs or what you have". No channels, no timers, just a simple data structure (probably just a slice) which is trivial to throw in to a test. The actual consumption of the data structure is also easier to follow at that point.

Reply:
Thanks for elaborating on the idea. I think actually trying it will help to see if it's worth it or not. I think if more logic needs to be packed into how IPs are handled then it may make sense. Another thing I'm thinking of is turning it into a map to dedupe IPs that have been in the queue for too long. And then we can just have a capacity limit on that map. Or maybe we don't even need that capacity limit with deduplication...